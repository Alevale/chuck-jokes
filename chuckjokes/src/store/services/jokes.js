export const getJokes = (amount) => {
    return fetch(`http://localhost:3000/api/v1/jokes/random/${amount || 10}`)
        .then(result => result.json())
        .then(items => {
            return items.data.value
        })
}
