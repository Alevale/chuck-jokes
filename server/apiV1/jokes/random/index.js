'use strict';

const jokes = require('express').Router();
const axios = require('axios');

jokes.get('/:amount?', async (req, res) => {
    const { amount = 10 } = req.params;

    // NOTE: Get the list of jokes to retrieve 20190430:Alevale
    const { data } = await axios.get(`http://api.icndb.com/jokes/random/${amount}`);

    // Return the User Object
    res.status( 200 ).json({
        data
    });
});

// Export the jokes
module.exports = jokes;

