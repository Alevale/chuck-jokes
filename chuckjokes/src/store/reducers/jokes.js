import {getJokes} from '../services/jokes'
import uniqBy from 'lodash/uniqBy'

const initState = {
    expanded: ['panel1', 'panel2'],
    jokes: [],
    loading: true,
    intervalActive: undefined,
    favoriteJokes: [],
    page: 0,
}

export const LOAD_JOKES = 'LOAD_JOKES'
export const ADD_FAVORITE_JOKE = 'ADD_FAVORITE_JOKE'
export const REPLACE_FAVORITE_JOKE = 'REPLACE_FAVORITE_JOKE'
export const TOGGLE_PANEL = 'TOGGLE_PANEL'
export const REMOVE_FAVORITE_JOKE = 'REMOVE_FAVORITE_JOKE'
export const SET_ACTIVE_INTEVAL = 'SET_ACTIVE_INTEVAL'
export const REMOVE_ACTIVE_INTEVAL = 'REMOVE_ACTIVE_INTEVAL'
export const PAGE_UP = 'PAGE_UP'
export const PAGE_DOWN = 'PAGE_DOWN'

export const loadJokes = (jokes) => ({ type: LOAD_JOKES , payload: jokes })
export const addJoke = (joke) => ({ type: ADD_FAVORITE_JOKE , payload: joke })
export const replaceFavoriteJokes = (jokes) => ({ type: REPLACE_FAVORITE_JOKE , payload: jokes })
export const removeJoke = (joke) => ({ type: REMOVE_FAVORITE_JOKE, payload: joke })
export const togglePanels = (panel) => ({ type: TOGGLE_PANEL, payload: panel })
export const setActiveInterval = (interval) => ({ type: SET_ACTIVE_INTEVAL, payload: interval })
export const removeActiveInterval = () => ({ type: REMOVE_ACTIVE_INTEVAL })
export const pageUp = (page) => ({ type: PAGE_UP , payload: page })
export const pageDown = (page) => ({ type: PAGE_DOWN , payload: page })

export const fetchJokes = () => {
    return (dispatch) => {
        getJokes()
            .then(jokes => dispatch(loadJokes(jokes)))
    }
}

export const buildFavoritesListFromServer = (userItems) => {
    return (dispatch, getState) => {
        const {favoriteJokes} = getState().joke
        if (userItems.result && userItems.result.joke && userItems.result.joke.favoriteJokes) {
            // NOTE: Do this so that if there were favoriteJokes they are all added 20190506:Alevale
            const favs = favoriteJokes.concat(userItems.result.joke.favoriteJokes)
            dispatch(replaceFavoriteJokes(uniqBy(favs, 'id')))
        }
    }
}

export const addFavoriteJokeOnIntervals = () => {
    return (dispatch, getState) => {
        const { intervalActive } = getState().joke
        if (!intervalActive) {
            const tempInterval = setInterval(() => {
                const { intervalActive, favoriteJokes } = getState().joke
                if (favoriteJokes.length >= 10) {
                    clearInterval(intervalActive)
                } else {
                    getJokes(1)
                        .then(joke => dispatch(addJoke(joke)))
                }
            }, 5000)
            dispatch(setActiveInterval(tempInterval))
            getJokes(1)
                .then(joke => dispatch(addJoke(joke)))
        } else {
            clearInterval(intervalActive)
            dispatch(removeActiveInterval())
        }
    }
}

const addToFavorites = (state, joke) => {
    const { jokes, favoriteJokes } = state
    return {
        ...state,
        jokes: jokes.filter((j) => j.id !== joke.id),
        favoriteJokes: uniqBy(favoriteJokes.concat(joke), 'id')
    }
}

const removeFromFavorites = (state, joke) => {
    const { jokes, favoriteJokes } = state
    return {
        ...state,
        favoriteJokes: uniqBy(favoriteJokes.filter((j) => j.id !== joke.id), 'id'),
        jokes: uniqBy(jokes.concat(joke), 'id')
    }
}

const togglePanel = (state, panel) => {
    const { expanded } = state
    return {
        ...state,
        expanded: expanded.includes(panel) ? expanded.filter((i) => i !== panel) : expanded.concat(panel)
    }
}

export default (state = initState, action) => {
    switch (action.type) {
        case LOAD_JOKES:
            return { ...state, jokes: action.payload, loading: false }
        case ADD_FAVORITE_JOKE:
            return addToFavorites(state, action.payload)
        case TOGGLE_PANEL:
            return togglePanel(state, action.payload)
        case REPLACE_FAVORITE_JOKE:
            return {...state, favoriteJokes: action.payload}
        case REMOVE_FAVORITE_JOKE:
            return removeFromFavorites(state, action.payload)
        case SET_ACTIVE_INTEVAL:
            return {...state, intervalActive: action.payload}
        case REMOVE_ACTIVE_INTEVAL:
            return {...state, intervalActive: undefined}
        case PAGE_UP:
            return { ...state, page: state.page + 1 }
        case PAGE_DOWN:
            return { ...state, page: state.page - 1 }
        default:
            return state
    }
}
