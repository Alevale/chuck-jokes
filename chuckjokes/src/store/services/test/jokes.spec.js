import {getJokes} from '../jokes'
import {joke1} from './examples'

describe('Joke Service', () => {
    it('fetches data from the server and converts it to json', done => {
        const mockSuccessResponse = {
            data: {
                value: { joke1 }
            }
        }
        const mockJsonPromise = Promise.resolve(mockSuccessResponse)
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise
        })
        jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise)

        const jokes = getJokes(3)

        expect(global.fetch).toMatchSnapshot()

        process.nextTick(() => {
            expect(jokes).toMatchSnapshot()

            global.fetch.mockClear()
            done()
        })
    })

    it('fetches 10 jokes by default', done => {
        const mockSuccessResponse = {
            data: {
                value: { joke1 }
            }
        }
        const mockJsonPromise = Promise.resolve(mockSuccessResponse)
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise
        })
        jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise)

        const jokes = getJokes()

        expect(global.fetch).toMatchSnapshot()

        process.nextTick(() => {
            expect(jokes).toMatchSnapshot()

            global.fetch.mockClear()
            done()
        })
    })

    it('fetches an unlimited amount of jokes', done => {
        const mockSuccessResponse = {
            data: {
                value: { joke1 }
            }
        }
        const mockJsonPromise = Promise.resolve(mockSuccessResponse)
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise
        })
        jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise)

        const jokes = getJokes(999)

        expect(global.fetch).toMatchSnapshot()

        process.nextTick(() => {
            expect(jokes).toMatchSnapshot()

            global.fetch.mockClear()
            done()
        })
    })
})
