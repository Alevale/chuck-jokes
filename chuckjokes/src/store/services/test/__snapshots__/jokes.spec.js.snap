// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Joke Service fetches 10 jokes by default 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://localhost:3000/api/v1/jokes/random/10",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`Joke Service fetches 10 jokes by default 2`] = `Promise {}`;

exports[`Joke Service fetches an unlimited amount of jokes 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://localhost:3000/api/v1/jokes/random/999",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`Joke Service fetches an unlimited amount of jokes 2`] = `Promise {}`;

exports[`Joke Service fetches data from the server and converts it to json 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://localhost:3000/api/v1/jokes/random/3",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`Joke Service fetches data from the server and converts it to json 2`] = `Promise {}`;
